internal/luks2/cryptsetup.go

```diff
@@ -0,0 +1,202 @@
+// the supplied benchmark time.
+func AddKey(devicePath string, existingKey, key []byte, kdfTime time.Duration) error {
+	fifoPath, cleanupFifo, err := mkFifo()
+	if err != nil {
+		return xerrors.Errorf("cannot create FIFO for passing existing key to cryptsetup: %w", err)
+	}
+	defer cleanupFifo()
+
+	args := []string{
+		// add a new key
+		"luksAddKey",
+		// LUKS2 only
+		"--type", "luks2",
+		// read existing key from named pipe
+		"--key-file", fifoPath,
+		// use argon2i as the KDF
+		"--pbkdf", "argon2i"}
+	if kdfTime != 0 {
+		// set the KDF benchmark time
+		args = append(args, "--iter-time", strconv.FormatUint(uint64(kdfTime/time.Millisecond), 10))
+	}
+	args = append(args,
+		// container to add key to
+		devicePath,
+		// read new key from stdin
+		"-")
+	cmd := exec.Command("cryptsetup", args...)
+	cmd.Stdin = bytes.NewReader(key)
+
+	var b bytes.Buffer
+	cmd.Stdout = &b
+	cmd.Stderr = &b
+
+	if err := cmd.Start(); err != nil {
+		return xerrors.Errorf("cannot start cryptsetup: %w", err)
+	}
+
+	f, err := os.OpenFile(fifoPath, os.O_WRONLY, 0)
+	if err != nil {
+		// If we fail to open the write end, the read end will be blocked in open()
+		cmd.Process.Kill()
+		return xerrors.Errorf("cannot open FIFO for passing existing key to cryptsetup: %w", err)
+	}
+
+	if _, err := f.Write(existingKey); err != nil {
+		f.Close()
+		// The read end is open and blocked inside read(). Closing our write end will result in the
+		// read end returning 0 bytes (EOF) and exitting cleanly.
+		cmd.Wait()
+		return xerrors.Errorf("cannot pass existing key to cryptsetup: %w", err)
+	}
+
+	f.Close()
+	if err := cmd.Wait(); err != nil {
+		return osutil.OutputErr(b.Bytes(), err)
+	}
+	return nil
+}
+
+// ImportToken imports the supplied token in to the JSON metadata area of the specified LUKS2 container.
+func ImportToken(devicePath string, token *Token) error {
+	tokenJSON, err := json.Marshal(token)
+	if err != nil {
+		return xerrors.Errorf("cannot serialize token: %w", err)
+	}
+	cmd := exec.Command("cryptsetup", "token", "import", devicePath)
+	cmd.Stdin = bytes.NewReader(tokenJSON)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return osutil.OutputErr(output, err)
+	}
+	return nil
+}
+
+// RemoveToken removes the token with the supplied ID from the JSON metadata area of the specified
+// LUKS2 container.
+func RemoveToken(devicePath string, id int) error {
+	cmd := exec.Command("cryptsetup", "token", "remove", "--token-id", strconv.Itoa(id), devicePath)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return osutil.OutputErr(output, err)
+	}
+	return nil
+}
+
+// KillSlot erases the keyslot with the supplied slot number from the specified LUKS2 container.
+// Note that a valid key for a remaining keyslot must be supplied, in order to prevent the last
+// keyslot from being erased.
+func KillSlot(devicePath string, slot int, key []byte) error {
```

Comment:
shouldn't this be called KillKeyslot for consistency ?

Reply:
I don't mind, although the function name here matches the cryptsetup command name

Reply:
I see, I missed that in my first pass (as I said I didn't quite do a full review yet).  Are there slots which are not keyslots? otherwise for consistency we could instead change SetKeyslotPriority to SetSlotPriority?

I don't have  a strong preference and it's a nitpick but is a bit not ideal to have both Slot and Keyslot interchangebly in the function names.

Reply:
I've renamed SetKeyslotPriority to SetSlotPriority (and the corresponding enum type as well). It's unfortunate that the cryptsetup interface uses both terms interchangeably.